### PR TITLE
First iteration of performance tweaks

### PR DESCRIPTION
**Pull request by mrmlnc**
### Replacing the placeholders

`string.replace` 2-2.5x faster than `vm.runInContext`, so we can use `string.replace` for simple expression (without dynamic calculations).

If the expression has something besides letters and numbers - use **VM** otherwise **replace**.

#### Benchmark [150x `{{ key }}`]

| runInContext | runInContext + replace | faster |
|--------------|------------------------|--------|
| 234 op/s     | 530 op/s               | 2.3x   |

### Loop statement parser

Use new solution for parsing loop statement

#### Benchmark
  
| Current solution | New solution | faster |
|------------------|--------------|--------|
| 196 450 op/s      | 1 798 479 op/s | 9.2x   |

#### Benchmark [20x `if - else if -else if - else`]

| Current solution | New solution | faster |
|------------------|--------------|--------|
| 322 op/s      | 349 op/s | 1.08x   |

## Types of Changes

Fully compatible with the current version.

### CODE REVIEW

**Comment by mrmlnc** [comment]
Don't merge this now, please. I want to add some tests.

File: .travis.yml [hunk 1]
```diff
@@ -2,5 +2,7 @@ language: node_js
 sudo: false
```
Comment (michael-ciniawsky) [inline .travis.yml]:
```yaml
language: node_js
node_js:
  - node
  - 6
  - 4
cache:
  directories:
    - node_modules
before_script:
  - npm i
script:
  - npm test
after_success:
- './node_modules/.bin/nyc report --reporter=text-lcov | ./node_modules/.bin/coveralls'
notifications:
  email: false

```

File: lib/placeholders.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    '"': '&quot;',
+    '\'': '&#039;'
+  })[match])
+};
+
+module.exports = function replace (input, ctx, settings) {
+  // Since we are matching multiple sets of delimiters, we need to run a loop
+  // here to match each one.
+  for (let i = 0; i < settings.length; i++) {
+    const matchs = input.match(settings[i].regexp)
```
Comment (michael-ciniawsky) [inline lib/placeholders.js]: typo ? matchs = matches

File: lib/index.js [hunk 3]
```diff
@@ -1,26 +1,57 @@
...
-  // conditional and loop options
-  conditionals = options.conditionalTags || ['if', 'elseif', 'else']
-  scopes = options.scopeTags || ['scope']
-  loops = options.loopTags || ['each']
+module.exports = function PostHTMLExpressions (_options) {
+  // set default options
+  options = Object.assign({
+    delimiters: ['{{', '}}'],
+    unescapeDelimiters: ['{{{', '}}}'],
+    conditionalTags: ['if', 'elseif', 'else'],
```
Comment (michael-ciniawsky) [inline lib/index.js]: Are options for setting the conditional-, loop-, scope tags really necessary? what if someone tries to add e.g options.loopTags = ['each', 'foreach', 'forof', 'forin'...] expecting them to work like the normally do in JS ? :) To reduce eventual confusion and configuration imo, we should maybe consider removing them from options. cc @voischev @posthtml/collaborators

File: lib/placeholders.js (hunk 2 above)
Comment (mrmlnc) [inline lib/placeholders.js]: Yeap, IntelliSense 🗡

File: lib/index.js (hunk 3 above)
Comment (michael-ciniawsky) [inline lib/index.js]: Ok I see, that's 👍  for dropping I would say :). Fixed tags for logical expressions, matching the JS naming conventions && logic when possible, is a valuable thing. Everyone can still use his/her's favourite delimiter flavour of choice. But not my decision alone, lets wait for input of others :)

**Comment by michael-ciniawsky** [comment]
@mrmlnc Whats your opinion on a `<switch>` tag, mimicking a switch statement?
```js
<switch>
  <case>
  <case>
  <case>
</switch>
```
likely this one is total rubbish 😛 , anyways...

`<js>Like a script tag but gets executed in Context</js>`

**Comment by mrmlnc** [comment]
@michael-ciniawsky,

I want to talk about it soon. Because all these changes do not solve the problem (a needle in a haystack), which was announced in #26.

**Comment by michael-ciniawsky** [comment]
Are there no known faster implementations of `merge` and `clonedeep` ? 😛

**Comment by mrmlnc** [comment]
@michael-ciniawsky,

Spoiler: even a simple `Object.assign({}, options)` kills performance from 500 op/s to 6-10 op/s.

**Comment by michael-ciniawsky** [comment]
uhhhh... 😟  I'm an extensive `Object.assign`  user

**Comment by mrmlnc** [comment]
@michael-ciniawsky,

Ready to merge. Please look again 😄

**Comment by michael-ciniawsky** [comment]
@mrmlnc I make few chore/docs tweaks later today and publish new under `posthtml-expressions`

File: lib/index.js [hunk 4]
```diff
@@ -278,7 +276,7 @@ function executeLoop (loopParams, p1, p2, node) {
  */
 function executeScoped (scopedLocals, node) {
   // merge nondestructively into existing locals
-  const scopedOptions = merge(cloneDeep(options), { locals: scopedLocals })
+  const scopedOptions = merge(cloneDeep(data), { locals: scopedLocals })
```
Comment (voischev) [inline lib/index.js]: 🤔

File: lib/placeholders.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+'use strict'
+
+const vm = require('vm')
+
+function escapeHtmlCharacters (unsafe) {
+  return unsafe.replace(/[&<>"']/g, (match) => ({
```
Comment (voischev) [inline lib/placeholders.js]: Object and RE should be cached ;)

File: lib/index.js (hunk 4 above)
Comment (mrmlnc) [inline lib/index.js]: `globals` / `globalLocals`?

File: lib/index.js [hunk 6]
```diff
@@ -1,29 +1,64 @@
...
-  delimiterRegex = new RegExp(`.*${delimiters[0]}(.*)${delimiters[1]}.*`, 'g')
-  unescapeDelimiterRegex = new RegExp(`.*${unescapeDelimiters[0]}(.*)${unescapeDelimiters[1]}.*`, 'g')
+let delimitersSettings = []
+let conditionals, loops, scopes, data
 
-  // identification for delimiter options, for the parser
-  delimiters.push('escaped')
-  unescapeDelimiters.push('unescaped')
+function escapeRegexpString (input) {
+  return input.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
```
Comment (mrmlnc) [inline lib/index.js]: Will be resolve by `performance-2`

File: lib/placeholders.js (hunk 5 above)
Comment (mrmlnc) [inline lib/placeholders.js]: Will be resolve by `performance-2`

File: lib/index.js (hunk 4 above)
Comment (voischev) [inline lib/index.js]: `merge`/`cloneDeep` :)

**Pull request merged**